book/src/todo_app_1.md

```diff
@@ -0,0 +1,231 @@
+- a string holding the content.
+
+<span class="filename">Filename: listings/todo_app/1/todo_object/mod.rs</span>
+
+```rust
+{{#rustdoc_include ../listings/todo_app/1/todo_object/mod.rs:glib_wrapper_and_new}}
+# // Please ignore this line
+# // It is only there to make mdbook happy
+# fn main() {}
+```
+
+Unlike the lists chapter, the state is stored in a struct rather than in individual members of `imp::TodoObject`.
+
+<span class="filename">Filename: listings/todo_app/1/todo_object/mod.rs</span>
+
+```rust
+{{#rustdoc_include ../listings/todo_app/1/todo_object/mod.rs:todo_data}}
+# // Please ignore this line
+# // It is only there to make mdbook happy
+# fn main() {}
+```
+
+Not only that, we see that it is also wrapped in a [`Rc`](https://doc.rust-lang.org/std/rc/struct.Rc.html).
+This is not really relevant for now, but will be very convenient when we want to save the state in one of the following chapters.
+Exposing `completed` and `content` as properties does not become much different that way, so we will not discuss it further.
+If you are curious, you can press on the small eye symbol on the top right of the code snippet to read the implementation.
+
+<span class="filename">Filename: listings/todo_app/1/todo_object/imp.rs</span>
+
+```rust
+{{#rustdoc_include ../listings/todo_app/1/todo_object/imp.rs:struct_and_subclass}}
+# // Please ignore this line
+# // It is only there to make mdbook happy
+# fn main() {}
+```
+
+Let us move on to the individual tasks.
+In the original mockup, we already determined that the row of a task should look like this:
+
+
+<div style="text-align:center"><img src="img/todo_row.png" /></div>
+
+Again, we describe the mockup with a composite template.
+
+
+<span class="filename">Filename: listings/todo_app/1/todo_row/todo_row.ui</span>
+
+```xml
+{{#rustdoc_include ../listings/todo_app/1/todo_row/todo_row.ui}}
+```
+
+First we [derive](https://docs.gtk.org/gtk4/class.Box.html#hierarchy) `TodoRow` from `gtk:Box`.
+
+<span class="filename">Filename: listings/todo_app/1/todo_row/mod.rs</span>
+
+```rust
+{{#rustdoc_include ../listings/todo_app/1/todo_row/mod.rs:glib_wrapper}}
+# // Please ignore this line
+# // It is only there to make mdbook happy
+# fn main() {}
+```
+
+In `imp::TodoRow`, we hold references to `completed_button` and `content_label`.
+We also store a mutable vector of bindings.
+Why we need that will become clear as soon as we get to bind the state of `TodoObject` to the corresponding `TodoRow`.
+
+
+<span class="filename">Filename: listings/todo_app/1/todo_row/imp.rs</span>
+
+```rust
+{{#rustdoc_include ../listings/todo_app/1/todo_row/imp.rs:struct_and_subclass}}
+# // Please ignore this line
+# // It is only there to make mdbook happy
+# fn main() {}
+```
+
+Now we can bring everything together.
+We override the `imp::Window::constructed` in order to ensure that everything will be setup immediately after the parent is constructed.
```

Comment:
> immediately after the parent is constructed

This particular part is a bit confusing and not really relevant I think, the point is that we set up window contents at the time of its construction